### DISCUSSION TITLE
DAI 2023-2024 - Java IOs benchmarking - Lionel Pollien

### DISCUSSION THREAD

**Discussion by poLionel** [heig-vd-dai-course/heig-vd-dai-course#97] [Show and tell]
https://github.com/poLionel/javaIosPracticalContent

# <u>Results</u>

Write in a binary file without | with buffer :
    
          1 byte  :     1 ms |    1 ms
       1024 bytes :    21 ms |    4 ms
    1048576 bytes :  3492 ms |  411 ms
    5242880 bytes : 17071 ms | 1429 ms

Read in a binary file without | with buffer : 

          1 byte  :    0 ms |   0 ms
       1024 bytes :    4 ms |   1 ms
    1048576 bytes : 1835 ms |  45 ms
    5242880 bytes : 8813 ms | 157 ms

Write in text file without | with buffer :

          1 byte  :    1 ms |    1 ms
       1024 bytes :   12 ms |    4 ms
    1048576 bytes :  382 ms |  324 ms
    5242880 bytes : 1511 ms | 1428 ms

Read in text file without | with buffer :

          1 byte  :   1 ms |   0 ms
       1024 bytes :   7 ms |   1 ms
    1048576 bytes :  74 ms |  35 ms
    5242880 bytes : 215 ms | 138 ms

### <u>Questions</u>

1) Which type of stream is the most efficient for each use case?

   The buffered stream are the most efficient

2) Why is it more efficient than the other types of streams?

   A buffered reader is more efficient than another reade cause it is faster to read data from memory instead of reading data from a disk or a network. All reading is done from the buffer. the disk or the network needs to be accessed occasionally to fill the buffer.

3) What is the difference between binary data and text data?

   A binary file and a text file both contain data as a series of bits. The difference is that for text it reprensents characters and for binary it reprensents custom data.

4) What is a character encoding?

   According to wikipedia, "character encoding is the process of assigning numbers to graphical characters". So this means that character encoding is a human character to bits conversion or vice versa and it allows the computer to interpret digital data into letters, numbers and symbols.

5) Why is this methodology important?

   First as said in question 4, character encoding allows computer to interpret our letters, numbers and symbols. Since we have different languages and different alphabets, it is usefull to create and process text in a multiple languages. Character encoding ensures the integrity of data as well when it is transmitted or stored digitally. Using the wrong encoding may results in data corruption.

**Comment by ludelafo** [comment]
Thank you!

Regarding the methodology, we were more referencing to the scientific methodology to benchmark our applications.

Using a systematic and proven methodology with tangible metrics, we can use this scientific methodology to benchmark our applications to ensure the performance are still good over time for the use cases we are managing.

Everything else seems good to me!
